ReSharper.FSharp/src/FSharp/FSharp.Psi.Features/src/AICore/FSharpFileSummarizer.fs

```diff
@@ -1,36 +1,271 @@
 namespace JetBrains.ReSharper.Plugins.FSharp.Psi.Features.AICore
 
+open System
+open System.Collections.Generic
+open System.Text
 open JetBrains.ReSharper.Plugins.FSharp.Psi
+open JetBrains.ReSharper.Plugins.FSharp.Psi.Features.Util.FcsTypeUtil
+open JetBrains.ReSharper.Plugins.FSharp.Psi.Impl
 open JetBrains.ReSharper.Plugins.FSharp.Psi.Tree
+open JetBrains.ReSharper.Plugins.FSharp.Util
 open JetBrains.ReSharper.Psi
+open JetBrains.ReSharper.Psi.Tree
 open JetBrains.Rider.Backend.Features.AICore.Summarization
 open JetBrains.ReSharper.Psi.Files
+open FSharp.Compiler.Symbols
+open JetBrains.Util
+
+type private SummarizerContext() =
+    let scopes =  Stack<string>()
+    let summary = StringBuilder()
+
+    member x.CurrentScope = scopes.TryPeek()
+    member x.Summary = summary.ToString()
+
+    member x.AddEntity(node: ITreeNode, representation: string) =
+        let range = node.GetDocumentRange()
+        let startLine, endLine =
+            range.StartOffset.ToDocumentCoords().Line.Plus1(),
+            range.EndOffset.ToDocumentCoords().Line.Plus1()
+
+        [x.CurrentScope; representation; $"({startLine}-{endLine})"]
+        |> Seq.filter _.IsNotEmpty()
+        |> String.concat " "
+        |> summary.AppendLine
+        |> ignore
+
+    member x.OpenScope(node: ITreeNode, scopeRepresentation: string, ?declarationInfo: string) =
+        let currentScope = x.CurrentScope
+        let scopeRepresentation =
+            match node with
+            | :? IModuleDeclaration -> scopeRepresentation
+            | _ -> if isNull currentScope then scopeRepresentation else " " + scopeRepresentation
+
+        scopes.Push(currentScope + scopeRepresentation)
+        x.AddEntity(node, defaultArg declarationInfo "")
+        { new IDisposable with member this.Dispose() = scopes.Pop() |> ignore }
+
+type private FileSummarizerVisitor() =
+    inherit TreeNodeVisitor<SummarizerContext>()
+
+    static let displayContext = emptyDisplayContext
+
+    static let formatFcsSymbolType (fcsSymbol: FSharpSymbol) =
```

Comment:
Could you explain why `static` was required, please?

Reply:
removed